microscopium/preprocess.py

```diff
@@ -808,6 +808,163 @@ def montage(ims, order=None):
+    order : array-like of int, shape (M, N), optional
+        The order of the stitching, with each entry referring
+        to the index of file in the fns array.
+    rows : int, optional
+        The number of rows in the input images. Default 512.
+    cols : int, optional
+        The number of cols in the input images. Default 512.
+
+    Returns
+    -------
+    mask : array of bool, shape (P, Q)
+        A binary mask where False denotes a missing field.
+    """
+    if order is None:
+        from .screens import cellomics
+        order = cellomics.SPIRAL_CLOCKWISE_RIGHT_25
+    order = np.atleast_2d(order)
+    mrows, mcols = order.shape
+
+    mask = np.ones((rows * mrows, cols * mcols),
+                   dtype=bool)
+
+    for i in range(mrows):
+        for j in range(mcols):
+            if order[i, j] in missing:
+                mask[rows*i:rows*(i+1), cols*j:cols*(j+1)] = False
+
+    return mask
+
+
+def montage_with_missing(fns, order=None):
+    """Montage a list of images, replacing missing fields with dummy values.
+
+    The methods `montage` and `montage_stream` assume that image filenames
+    and image iterators passed to it are complete, and include the full set
+    images belonging to the well. Some screens have missing fields,
+    so this function can be used to montage together images with missing
+    fields. Missing fields are replaced with 0 values.
+
+    Missing fields are determined from the information in the image
+    file name. See 'find_missing_fields'
+
+    Parameters
+    ----------
+    fns : list of str
+        The list of filenames to montage.
+    order : array-like of int, shape (M, N), optional
+        The order of the stitching, with each entry referring
+        to the index of file in the fns array.
+        Default cellomics.SPIRAL_CLOCKWISE_RIGHT_25
+
+    Returns
+    -------
+    montaged : array-like, shape (P, Q)
+        The montaged image.
+    mask : array of bool, shape (P, Q)
+        A binary mask, where entries with taking the value of
+        False represent missing fields in the montaged image.
+    missing : int
+        The number of fields that were found to be missing in the
+        input list of filenames. This is useful for normalising
+        features that depend on the entirety of the montaged image
+        (e.g. count of objects).
+    """
+    if order is None:
+        from .screens import cellomics
+        order = cellomics.SPIRAL_CLOCKWISE_RIGHT_25
+    order = np.atleast_2d(order)
+    mrows, mcols = order.shape
+
+    # get width & height of first image. the rest of the images
+    # are assumed to be of the same shape
+    im0 = io.imread(fns[0])
+    rows, cols = im0.shape[:2]
+
+    # find which fields are missing
+    missing = find_missing_fields(fns, order)
+
+    # insert None value to list of files when fields missing
+    _fns = fns[:]  # create copy of list to avoid referencing problems
+    for i in missing:
+        _fns.insert(i, None)
```

Comment:
Oh yeah, of course option 2 is better. I only left 1 in there cos I'd already typed it in when I thought of 2 and didn't want the effort to go to waste. =P